MKL-DNN integration: request for reviews

ykim362 opened the pull request:
**This PR is a beta version for the code reviews and experiments. There are several known issues which are being debugged.**

**If this version is built with 'USE_MKL2017' and 'USE_MKL2017_EXPERIMENTAL' flag, it will provide the same functionalities and performance as the current MKLML release. If this is built with 'USE_MKLDNN' flag, it will go through the new code pass (MKL-DNN integration).**


# MKL-DNN
A new open-source deep learning library providing IA optimized DNN kernels.
https://github.com/01org/mkl-dnn

# Advantages
## More functionalities
New functionalities will be mainly added to MKL-DNN rather than MKLML library.
Below are two examples.
1. Fused RNN cell (To be added)
2. int8 inference (To be added)

## Performance optimization
As of Sep. 18 2017. 
* all units are (img/sec)
Alexnet Inference (BS:256): 1474 (MKLML) --> 1568 (MKL-DNN)
inception-bn inference (BS:32): 454 (MKLML) --> 483 (MKL-DNN)
on Skylake 20-core machine (6148)
Resnet 50 inference (BS: 32): 99 (MKLML) --> 116 (MKL-DNN)
on KNL 7250

# Known issues
- Convergence (resnet training)

# Contributors for this PR
@ashokei @karkadad @louisfeng @adstraw

piiswrong commented on src/operator/mkl/mkldnn_memory.cc:
```diff
@@ -0,0 +1,285 @@
+/*******************************************************************************
+* Copyright 2016-2017 Intel Corporation
```
I'm not sure if Apache allows this.
@smarthi

piiswrong commented:
what's the issue with resnet convergence?

szha commented on prepare_mkl.sh:
```diff
@@ -115,7 +115,7 @@ if [ -z $MKLROOT ]; then
 # Check what MKL lib we have in MKLROOT
-if [ -z `find $MKLROOT -name libmklml_gnu.so -o -name libmklml.dylib -print -quit` ]; then
+if [ -z `find $MKLROOT -name libmklml_gnu.so -print -quit` ]; then
   USE_MKLML=0
```
I enabled mac mkl a while back. What's wrong with this?

szha replied:
Also, why is libmklml_gnu listed here while in most cases the linked library is libmklml_intel?

ykim362 replied:
This change makes USE_MKLML to 0 on linux machines.

szha replied:
Not at the price of breaking mac please.

ykim362 replied:
@szha This new change from @ashokei would resolve the problem.

ykim362 commented:
@piiswrong MKL-DNN doesn't converge with Resnet. It converges with mnist. I am looking into it to find a root cause. MKLML works fine for both resnet and mnist.

szha commented on src/operator/activation.cc:
```diff
@@ -29,12 +29,30 @@
+    case mshadow::kInt8:
+    case mshadow::kInt32:
+    case mshadow::kUint8:
+      return new MKLDNNReluOp<cpu, float>();
```
Should it be allowed to turn on both MKLDNN and MKL2017? MKLRelu below supports double which isn't supported by mkldnn here.

ykim362 replied:
This is missing. I will add double for MKLDNN.

ykim362 replied:
Either one of MKLML(2017) or MKLDNN should be used. They cannot be used together for now.

szha commented on src/operator/concat.cc:
```diff
@@ -50,6 +55,18 @@ Operator* CreateOp<cpu>(ConcatParam param, int dtype) {
+  if ((1 == param.dim) && (param.num_args > 1)) {
+    switch (dtype) {
+      case mshadow::kFloat32:
+        return new MKLDNNConcatOp<cpu, float>(param);
```
Similar comment as above.

ykim362 replied:
This is missing. I will fix.

szha replied:
There are several other ops that have the same problem. I omitted them for brevity. Let me know once you fixed the types.

szha commented on src/operator/lrn.cc:
```diff
@@ -40,6 +46,9 @@ Operator* CreateOp<cpu>(LRNParam param, int dtype) {
   return new MKLLRNOp<cpu, float>(param);
 #endif
+#if MXNET_USE_MKLDNN == 1
+  return new MKLDNNLRNOp<cpu, float>(param);
```
If both MKL2017 and MKLDNN are on, this would become dead code.

ashokei replied:
As we do not support both flags at the same time, our MKLDNN build config will ensure MKL2017 is not be used.

szha commented on src/operator/mkl/mkl_conv-common-inl.h:
```diff
@@ -0,0 +1,82 @@
+    stride_w_,
+    stride_h_;
+  int group_,
+    num_,
```
What's num?

ykim362 replied:
This means batch size.

szha replied:
Would you rename it?

ykim362 replied:
Yes, I will do!

ykim362 replied:
This actually follows the notation used in MXNet's convolution. https://github.com/apache/incubator-mxnet/blob/master/src/operator/convolution-inl.h#L369

szha commented on src/operator/mkl/mkl_memory.h:
```diff
@@ -112,6 +117,10 @@ struct MKLMemHolder {
 };
+
+// bool compare_mkl_memholder(std::shared_ptr<MKLMemHolder> holder_a,
+//   std::shared_ptr<MKLMemHolder> holder_b);
```
Remove commented code.

szha commented:
General comment: I didn't see mac being supported on mkl-dnn page. It only says the software was validated on RHEL7. What does this change imply for mac users?

sbodenstein commented:
@szha: I asked the same thing, this was the response: https://github.com/01org/mkl-dnn/issues/10#issuecomment-323730811

piiswrong commented:
Does consistency tests pass for mkldnn operators?

szha commented:
@ykim362 could you add a job for testing mxnet with mkldnn in Jenkinsfile? Otherwise the tests don't reflect the change.

ykim362 commented:
@piiswrong Yes, MKL-DNN build passes 'test_operator.py' unit test same as MKLML build. So, we are doing several experiments.

@szha I will looks into it to replicate the MKLML case.

eric-haibin-lin commented on src/operator/batch_norm.cc:
```diff
@@ -334,6 +339,30 @@ Operator *CreateOp<cpu>(BatchNormParam param, const int dtype, const TShape& sha
+    }
+  }
+#define BATCHNORM_LOG_MKL_INFO() \
+  do { \
```
why does this need a while loop?

ashokei replied:
this is for macro definition,  so it expands/compiles correctly in all cases.

eric-haibin-lin commented on src/operator/mkl/mkldnn_base.cc:
```diff
@@ -0,0 +1,74 @@
+using namespace mkldnn;
+namespace mxnet {
+
+bool enableMKLDNNWarnGenerated() {
```
This doesn't follow the naming convention for functions

eric-haibin-lin commented on src/operator/mkl/mkl_memory.h:
```diff
@@ -64,12 +64,16 @@ struct MKLMemHolder {
   }
   void set_prv_descriptor(std::shared_ptr<PrvMemDescr> descriptor, bool same_data = false) {
-    head_ = HEAD_AT_PRV;
+    if (descriptor != nullptr)
```
`    if (descriptor != nullptr)      head_ = HEAD_AT_PRV;` 
or 
```
    if (descriptor != nullptr) {
      head_ = HEAD_AT_PRV;
    }

```

eric-haibin-lin commented on src/operator/mkl/mkl_memory.h:
```diff
@@ -46,7 +46,7 @@ struct PrvMemDescr {
 };
 
-#if MKL_EXPERIMENTAL == 1
+#if MKL_EXPERIMENTAL == 1 || MXNET_USE_MKLDNN == 1
```
what does `Prv` stand for?

ashokei replied:
primitive/private layout

eric-haibin-lin commented on src/operator/mkl/mkl_memory.h:
```diff
@@ -64,12 +64,16 @@ struct MKLMemHolder {
     b_eager_mode = eager_mode;
```
what is `eager_mode ` used for? Is there any document for this?

eric-haibin-lin commented on src/operator/mkl/mkldnn_base-inl.h:
```diff
@@ -0,0 +1,164 @@
+class CpuEngine {
+ public:
+    static CpuEngine & Instance() {
+        // I's thread-safe in C++11.
```
typo?

eric-haibin-lin commented on src/operator/mkl/mkldnn_batch_norm-inl.h:
```diff
@@ -0,0 +1,414 @@
+template<typename xpu, typename Dtype>
+class MKLDNNBatchNormOp : public Operator, public MKLDNNLayer<Dtype> {
+ public:
+  static int s_id_gen;
```
what is `s_id_gen` used for? Add brief comment?

ashokei replied:
mostly used for debugging, to uniquely identify a node in a graph.

eric-haibin-lin commented on src/operator/mkl/mkldnn_batch_norm-inl.h:
```diff
@@ -0,0 +1,414 @@
+    // name = name + std::to_string(m_id);
+    return name;
+  }
+  // static std::string getName() {
```
Please move unused code

eric-haibin-lin commented on src/operator/mkl/mkldnn_concat-inl.h:
```diff
@@ -0,0 +1,326 @@
+namespace mxnet {
+namespace op {
+
+template<typename xpu, typename Dtype>
```
This file doesn't have any documentation/comments. Please add them for better maintainability

eric-haibin-lin commented on src/operator/mkl/mkldnn_convolution-inl.h:
```diff
@@ -0,0 +1,491 @@
+  }
+ private:
+  void InitForward(const OpContext &ctx) {
+      // if (std::is_same<DType, double>::value)   NOT_IMPLEMENTED;
```
Is this implemented now?

eric-haibin-lin commented on src/operator/mkl/mkldnn_memory-inl.h:
```diff
@@ -0,0 +1,206 @@
+          _internal_ptr = reinterpret_cast<Dtype *>(_prv_memory->get_data_handle());
+          _internal_size = prv_size();
+        } else {
+          VLOG(1) << "Set NULL Prv Memory";
```
When will this log be printed

ashokei replied:
it seems we currently do not use this function, it is there for future use-cases. if the caller of this function fails in creating mkldnn memory, it tries to print debug msg.

eric-haibin-lin commented on src/operator/mkl/mkldnn_memory.cc:
```diff
@@ -0,0 +1,285 @@
+
+
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Does this pass lint?

eric-haibin-lin commented:
How is this PR tested right now?

ykim362 commented:
@eric-haibin-lin We are adding CI test same as MKLML.

ashokei commented on include/mxnet/ndarray.h:
```diff
@@ -512,7 +512,7 @@ class NDArray {
+#if MKL_EXPERIMENTAL == 1 || MXNET_USE_MKLDNN == 1
     if (Mkl_mem_ != nullptr) {
       // convert prv to cpu
       Mkl_mem_->check_and_prv_to_cpu(ptr_->shandle.dptr);
```
we create layout conversion (cpu->prv) in each operator code transparently, this allows us to avoid layout conversions if the next (or preceding) operators are also mkldnn operators.  we convert back to cpu here:

https://github.com/apache/incubator-mxnet/pull/7931/files#diff-91ec751a29892f182cd05ff8de7c5755R135

check_and_prv_to_cpu can be used to convert to cpu layout, outside of the operator code, on an as needed basis.

ashokei commented on src/executor/attach_op_execs_pass.cc:
```diff
@@ -196,7 +201,7 @@ class FComputeExecutor : public StorageFallbackOpExecutor {
     PostFCompute(is_gpu);
-#if MKL_EXPERIMENTAL == 1
+#if MKL_EXPERIMENTAL == 1 || MXNET_USE_MKLDNN == 1
     mkl_tblobs_prv_to_cpu(in_data_);
```
we ensure we can convert back to cpu layout, it does not assume mutability or not.

szha commented:
@ykim362 It would be great if you could make sure that the bug in #8196 is resolved for at least the new mkl-dnn-based version before attempting to merge this change.

ykim362 commented:
@piiswrong We have fixed convergence issue in Resnet. There were some problems in Conv and Batch norm layers. Also, we have added some more optimizations to get more speed-ups. Now, MKL-DNN version is 15% faster than MKLML (MKL2017) for inference and training on average.

ykim362 commented:
@szha @piiswrong MKL-DNN doesn't support fp64(double) data type. Do you think this is an issue? The library team is more focusing on adding lower precisions.

szha commented:
@ykim362 could you verify if #8196 is fixed?

ykim362 commented:
@szha Sure, I am looking into it.

szha commented:
@ykim362 BTW is the fix in mklml_lnx_2018.0.20170908.tgz? Does it make sense to upgrade the library for mkl2017 use case? Many people are using MKL version (with MKL2017 and experimental on).

louisfeng commented on src/operator/mkl/mkldnn_elemwise_sum-inl.h:
```diff
@@ -0,0 +1,233 @@
+
+  if (req[0] == kNullOp) return;
+
+  Stream<xpu> *s = ctx.get_stream<xpu>();
```
This is true right now. We may need to use the stream when we try to support other tensor shapes beside nchw. I can remove it for now.

louisfeng commented on src/operator/mkl/mkldnn_elemwise_sum-inl.h:
```diff
@@ -0,0 +1,233 @@
+ * @param req
+ * @param out_data
+ */
+template<typename xpu, typename DType>
```
This is a good point. I was following the convention of other MKLDNN operators to be consistent. I can change this to cpu only.

louisfeng replied:
Also discussed with Young and Ashok, we would like to keep this template parameter for supporting future Intel devices. We may support devices other than traditional CPU in the future.

ykim362 commented:
@szha MKL-DNN also utilizes MKL2017. So, it would be useful to update MKL2017(2018) as well. And, the fix would be in the MXNet code. I am still investigating it.

louisfeng commented on src/operator/tensor/elemwise_binary_op_basic.cc:
```diff
@@ -23,11 +23,22 @@
+.set_attr<FInferStorageType>("FInferStorageType", 
+                              ElemwiseStorageType<2, 1, true, false, false>)       \
+.set_attr<FCompute>("FCompute<cpu>", MKLDNNElementWiseAddCompute<cpu>)             \
+.set_attr<FComputeEx>("FComputeEx<cpu>", ElemwiseBinaryOp::ComputeEx<cpu, mshadow::op::plus>)
```
Since FComputeEx is for NDArray inputs and all current MKLDNN operators are only supporting TBlobs and the main benefit of the MKLDNN elemwise sum operator comes from working with other MKLDNN operators. With the upcoming sparse tensor support, we may need to make some adjustments.

jesterhazy commented:
is there an update on this issue? we are keen to include MKL in an upcoming project.

ykim362 commented:
@piiswrong @szha @sbodenstein MKL-DNN now officially supports OSX. So, we don't need to worry about this issue.
https://github.com/01org/mkl-dnn/releases/tag/v0.11

sbodenstein commented:
@ykim362: thanks! I saw that. 

Btw: do you have any estimate for when this PR will be ready?

ykim362 commented:
@sbodenstein From my understanding, this PR is not going to be directly merged. It's going to be merged with another revision with sparse tensor (mkl storage). @piiswrong Is this correct?

sbodenstein commented:
@ykim362: do you know if bugs, like the resnet convergence bug, are still unsolved with v0.11 MKL-DNN?

leezu commented:
@sbodenstein MKL with v0.11 is quite buggy. I often got `inf` values during training for no obvious reason (i.e. training is stable without mkl / on GPU).
I'm not sure about v0.12 due to https://github.com/apache/incubator-mxnet/issues/8280 .

piiswrong commented:
closing since @zheng-da is making a new PR for this
